eth/beacon/state_machines/base.py

```diff
@@ -193,9 +222,562 @@ def active_state(self) -> ActiveState:
+            )
+        attestation = block.attestations[0]
+
+        is_proposer_attestation = (
+            attestation.shard_id == block_committees_info.proposer_shard_id and
+            attestation.slot == parent_block.slot_number and
+            has_voted(
+                attestation.attester_bitfield,
+                block_committees_info.proposer_index_in_committee
+            )
+        )
+        if not is_proposer_attestation:
+            raise ValidationError(
+                "Proposer of parent block should be one of the attesters in block.attestions[0]:\n"
+                "\tExpected: proposer index in committee: %d, shard_id: %d, slot: %d\n"
+                "\tFound: shard_id: %d, slot: %d, voted: %s" % (
+                    block_committees_info.proposer_index_in_committee,
+                    block_committees_info.proposer_shard_id,
+                    parent_block.slot_number,
+                    attestation.shard_id,
+                    attestation.slot,
+                    has_voted(
+                        attestation.attester_bitfield,
+                        block_committees_info.proposer_index_in_committee,
+                    ),
+                )
+            )
+
+    #
+    # Randao reveal validation
+    #
+    @classmethod
+    def validate_randao_reveal(cls) -> None:
+        # TODO: it's a stub
+        return
+
+    #
+    # Attestation validation
+    #
+    @classmethod
+    def validate_attestation(cls,
+                             block: BaseBeaconBlock,
+                             parent_block: BaseBeaconBlock,
+                             crystallized_state: CrystallizedState,
+                             active_state: ActiveState,
+                             attestation: 'AttestationRecord',
+                             chaindb: BaseBeaconChainDB,
+                             cycle_length: int) -> None:
+        """
+        Validate the given ``attestation``.
+
+        Raise ``ValidationError`` if it's invalid.
+        """
+        cls.validate_slot(
+            parent_block,
+            attestation,
+            cycle_length,
+        )
+
+        cls.validate_justified(
+            crystallized_state,
+            attestation,
+            chaindb,
+        )
+
+        attestation_indices = get_attestation_indices(
+            crystallized_state,
+            attestation,
+            cycle_length,
+        )
+
+        cls.validate_bitfield(attestation, attestation_indices)
+
+        cls.validate_version(crystallized_state, attestation)
+
+        parent_hashes = get_signed_parent_hashes(
+            active_state.recent_block_hashes,
+            block,
+            attestation,
+            cycle_length,
+        )
+        cls.validate_aggregate_sig(
+            crystallized_state,
+            attestation,
+            attestation_indices,
+            parent_hashes,
+        )
+
+    @classmethod
+    def validate_slot(cls,
```

Comment:
This looks like it can be a staticmethod which also suggests it could be extracted to a standalone to simplify the class (but still tacked onto the class as a `staticmethod` for convenience if you want).